src/lib/statistics/chunk_statistics/counting_quotient_filter.cpp

```diff
@@ -0,0 +1,163 @@
+  Assert(quotient_size > 0, "quotient size can not be zero.");
+  Assert(quotient_size + static_cast<uint8_t>(remainder_size) <= 64, "The hash length can not exceed 64 bits.");
+  Assert(remainder_size == 2 || remainder_size == 4 || remainder_size == 8 || remainder_size == 16
+    || remainder_size == 32, "remainder size must be 2, 4, 8, 16, or 32");
+
+  _remainder_size = remainder_size;
+  _number_of_slots = std::pow(2, quotient_size);
+  _hash_bits = quotient_size + remainder_size;
+
+if (remainder_size == 2) {
+    _quotient_filter = gqf2::quotient_filter();
+  } else if (remainder_size == 4) {
+    _quotient_filter = gqf4::quotient_filter();
+  } else if (remainder_size == 8) {
+    _quotient_filter = gqf8::quotient_filter();
+  } else if (remainder_size == 16) {
+    _quotient_filter = gqf16::quotient_filter();
+  } else if (remainder_size == 32) {
+    _quotient_filter = gqf32::quotient_filter();
+  }
+  boost::apply_visitor([&](auto& filter) {qf_init(&filter, _number_of_slots, _hash_bits, 0);}, _quotient_filter);
+}
+
+template <typename ElementType>
+CountingQuotientFilter<ElementType>::~CountingQuotientFilter() {
+  boost::apply_visitor([&](auto& filter) {qf_destroy(&filter);}, _quotient_filter);
+}
+
+template <typename ElementType>
+void CountingQuotientFilter<ElementType>::insert(ElementType element, uint64_t count) {
+  uint64_t bitmask = static_cast<uint64_t>(std::pow(2, _hash_bits)) - 1;
+  uint64_t hash = bitmask & _hash(element);
+  for (size_t i = 0; i < count; i++) {
+    boost::apply_visitor([&](auto& filter) {qf_insert(&filter, hash, 0, 1);}, _quotient_filter);
+  }
+}
+
+template <typename ElementType>
+void CountingQuotientFilter<ElementType>::insert(ElementType element) {
+  insert(element, 1);
+}
+
+template <typename ElementType>
+uint64_t CountingQuotientFilter<ElementType>::count_all_type(AllTypeVariant value) const {
+  DebugAssert(value.type() == typeid(ElementType), "Value does not have the same type as the filter elements");
+  return count(type_cast<ElementType>(value));
+}
+
+template <typename ElementType>
+bool CountingQuotientFilter<ElementType>::can_prune(const AllTypeVariant& value,
+    const PredicateCondition predicate_type) const {
+  DebugAssert(predicate_type == PredicateCondition::Equals, "CQF only supports equality predicates");
+  return count_all_type(value) == 0;
+}
+
+template <typename ElementType>
+uint64_t CountingQuotientFilter<ElementType>::count(ElementType element) const {
+  uint64_t bitmask = static_cast<uint64_t>(std::pow(2, _hash_bits)) - 1;
+  uint64_t hash = bitmask & _hash(element);
+  uint64_t count;
+  boost::apply_visitor([&](auto& filter) {count = qf_count_key_value(&filter, hash, 0);}, _quotient_filter);
+  return count;
+}
+
+/**
+* Computes the hash for a value.
+**/
+template <typename ElementType>
+uint64_t CountingQuotientFilter<ElementType>::_hash(ElementType value) const {
+  auto hash = murmur2<ElementType>(value, _seed);

```

Comment:
Didn't you say that the QCF bring their own hashing? Looks like we still need the murmur hashing, right?

Reply:
Yes the new version had their own hashing, but I couldn't integrate that version. Thats why I am using the old version.